### PR TITLE
Add Possibility to have multiple child joint 

### PR DESCRIPTION
**Pull request by CarlottaSartore**
This PR Adresses #21 

C.C. @GrmanRodriguez @AlexAntn @traversaro

### CODE REVIEW

File: urdfModifiers/core/fixedOffsetModifier.py [hunk 1]
```diff
@@ -156,15 +156,21 @@ def calculate_offsets(self):
...
-        child_joint_origin_z = (child_joint_origin[2] if child_joint_origin is not None else 0)
+        child_joint_origin_z = ({} if self.child_joint else {None:0.0})
+        if (self.child_joint is not None):  
+            for item in self.child_joint:
+                child_joint_origin = self.get_joint_origin(item)
+                child_joint_origin_z.update({item:(child_joint_origin[2] if child_joint_origin is not None else 0)})
 
         # Using formula: s_o = v_o - v_l * sign(j_o) / 2
-        parent_joint_offset = (link_visual_origin_z - link_length * math.copysign(1, child_joint_origin_z) / 2 if self.parent_joint else None)
+        parent_joint_offset = (link_visual_origin_z - link_length * math.copysign(1, float(child_joint_origin_z[list(child_joint_origin_z.keys())[0]])) / 2 if self.parent_joint else None)
```
Comment (CarlottaSartore) [inline urdfModifiers/core/fixedOffsetModifier.py]: I added to consider only the first joint for computing the offset with the parent, I do not think this is the smartest way to do it, does someone have any suggestion about it ? Thanks

File: tests/tests.py [hunk 2]
```diff
@@ -863,7 +863,7 @@ def test_calculate_offset_function(self):
         parent_joint_offset, child_joint_offset = modifier.calculate_offsets()
 
         self.assertEqual(parent_joint_offset, 0.5)
-        self.assertEqual(child_joint_offset, -0.5)
+        self.assertEqual(float(child_joint_offset[list(child_joint_offset.keys())[0]]), -0.5)
```
Comment (CarlottaSartore) [inline tests/tests.py]: The same comment as https://github.com/icub-tech-iit/urdf-modifiers/pull/24/files#r832063631 holds, in addition now child joint offset is a dict !

**Comment by traversaro** [comment]
Can you briefly describe how you are considering multiple joints in this PR? We can probably using the formalism used in https://github.com/icub-tech-iit/urdf-modifiers/issues/11 that is similar to the one used in the doc of the `fixedOffsetModifier` class.

**Comment by CarlottaSartore** [comment]
> Can you briefly describe how you are considering multiple joints in this PR? We can probably using the formalism used in #11 that is similar to the one used in the doc of the `fixedOffsetModifier` class.

I did not introduced any new formalism, basically now the computation for the child joint is applied to all the child joints of the link in the same fashion as per #11 

Let me know if it is not clear, I will then add a more formal description :)

**Comment by traversaro** [comment]
> Let me know if it is not clear, I will then add a more formal description :)

I was a bit confused by "in the same fashion as", I am not so able to go deep in the code, I will try to document it from the code.

If we have a `second_child_link` we need first of all to introduce in the notation, so taking the nice ascii art from https://github.com/icub-tech-iit/urdf-modifiers/blob/8a5b9bb6d0f76ac6706691695a2d88789f43a15c/urdfModifiers/core/fixedOffsetModifier.py#L15 we have:

~~~
  --------- +-----------+ -
  ^     ^   |           | ^
  |e_so |   |     ^     | |
  |     |   |     |     | |
  |-----|---|     o-->  | |------------------------
        |   |second_child |                       ^
        |   |           | |                       |
        |   |           | |                       |
    e_o |   |     ^     | |                       |
        |   |     |     | |                       |
        --- |     o-->  | |-----------            |
            |   child   | |          ^            |
            |           | | v_l      |            |
            |           | |          |            |
            |     o---- | |----      |            |
            |   visual  | |   ^      | j_o        | j_so
            |           | |   |      |            |
            |           | |   |      |            |
            |           | |   |      |            |
            |           | |   |      |            |
            |           | |   |      |            |
            |           | v   |      |            |
        --- +-----------+ -   | v_o  |            |
        ^         ^           |      |            |
    s_o |         |           |      |            |
        |         o-->        |      |            |
                parent
~~~

That correspond to the case:
~~~
  <link name="link">
    <visual>
      <origin xyz="0 0 v_o" rpy="0 -0 0"/>
      <geometry>
        <box size="0.1 0.1 v_l"/>
      </geometry>
    </visual>
  </link>
  <joint name="joint" type="continuous">
    <origin xyz="0 0 j_o" rpy="0 -0 0" />
    <axis xyz="0 1 0" />
    <parent link="link" />
    <child link="child_link" />
  </joint>
  <link name="child_link" />
  <joint name="second_joint" type="continuous">
    <origin xyz="0 0 j_so" rpy="0 -0 0" />
    <axis xyz="0 1 0" />
    <parent link="link" />
    <child link="second_child_link" />
  </joint>
  <link name="second_child_link" />
~~~

So, if we want to find a relation between the original quantities v_o, v_l and j_o and the resize quantities v_o', v_l' and j_o', we need to impose three constraints. In the original issue https://github.com/icub-tech-iit/urdf-modifiers/issues/11 we imposed (if K was the multiplier):
~~~
e_o' = e_o 
s_o' = s_o 
j_o' = Kj_o  
~~~
while in https://github.com/icub-tech-iit/urdf-modifiers/pull/20 it was then imposed a slighly different version, i.e. : 
~~~
e_o' = e_o 
s_o' = s_o 
v_l' = Kv_l  
~~~
(this was one of the reason I was confused, if the offset is now update following what was done in https://github.com/icub-tech-iit/urdf-modifiers/issues/11 this means that it changed w.r.t. to what was done in  https://github.com/icub-tech-iit/urdf-modifiers/pull/20?).

However, in both case the result is such that the offset do not change, i.e. `e_o' = e_o`
and `s_o' = s_o`

The resulting equations are (from the  code comments):
~~~
v_l' = Kv_l  
v_o' = s_o + v_l' * sign(j_o) / 2
j_o' = v_o' - e_o + v_l' * sign(j_o) / 2
~~~

**Comment by traversaro** [comment]
At this point, I tried to understand from the what is the equation used to update `j_so'`, but I got a bit lost in the code . It is possible that for congrunence with the rest of the code it is the equation imposing imposing `e_so' = e_so`, but to be honest I do not know if this is the case and that was the reason I asked.

**Comment by CarlottaSartore** [comment]
Thanks @traversaro ,   for the comment and the reasoning! It will for sure help in understanding possible bugs in the code :) 

Let me try to explain better what I implemented, starting from the fancy ascii you were also using: 
```
 --------- +-----------+ -
  ^     ^   |           | ^
  |e_o2 |   |     ^     | |
  |     |   |     |     | |
  |-----|---|     o-->  | |------------------------
  v     |   |second_child |                       ^
        |   |           | |                       |
        |   |           | |                       |
   e_o1 |   |     ^     | |                       |
        |   |     |     | |                       |
        v-- |     o-->  | |-----------            |
            |   child   | |          ^            |
            |           | | v_l      |            |
            |           | |          |            |
            |     o---- | |----      |            |
            |   visual  | |   ^      | j_o1        | j_o2
            |           | |   |      |            |
            |           | |   |      |            |
            |           | |   |      |            |
            |           | |   |      |            |
            |           | |   |      |            |
            |           | v   |      |            |
        --- +-----------+ -   | v_o  |            |
        ^         ^           |      |            |
    s_o |         |           |      |            |
        |         o-->        |      |            |
        v       parent        v      v            v
```

Basically what we want to compute is $e\_{i0}$ for each joint $i \in child$ in the case depicted $child = [1,2]$ and $s\_0$  which are the quantities that we want to keep constant throughout the scaling. 
Such quantities should then be constant, independently of the modification applied, and can be computed as: 

$$
\begin{equation}
e\_{o,i} = v\_o +v\_l \frac{sign(j\_{oi})}{2} - j\_{o,i}
\end{equation}
$$

**Code**
https://github.com/CarlottaSartore/urdf-modifiers/blob/37291f9e64a2071c4883b925a0002bb880d6ec10/urdfModifiers/core/fixedOffsetModifier.py#L171-L172

$$
\begin{equation}
s\_o = v\_o- v\_l \frac{sign(j\_{o,1})}{2}
\end{equation}
$$

**Code**
https://github.com/CarlottaSartore/urdf-modifiers/blob/37291f9e64a2071c4883b925a0002bb880d6ec10/urdfModifiers/core/fixedOffsetModifier.py#L166

**Note** for computing the parent offset, only one joint is taken into account, namely the first one. 

Given a length multiplier $k$ we can scale the component as follow: 

$$
\begin{equation}
v\_l^{'} = kv\_l
\end{equation}
$$
**Code:**


https://github.com/CarlottaSartore/urdf-modifiers/blob/37291f9e64a2071c4883b925a0002bb880d6ec10/urdfModifiers/core/fixedOffsetModifier.py#L194

$$ 
\begin{equation}
v\_{o}^{'} = s\_o+v\_l{'}\frac{sign(j\_{o,1})}{2}
\end{equation}
$$

**Code:**
https://github.com/CarlottaSartore/urdf-modifiers/blob/37291f9e64a2071c4883b925a0002bb880d6ec10/urdfModifiers/core/fixedOffsetModifier.py#L211

$$
j\_{o,i}^{'} = s\_o + v\_{l}^{'}sign({j\_{o,i}}) - e\_{o,i}
$$

**Code:**
https://github.com/CarlottaSartore/urdf-modifiers/blob/37291f9e64a2071c4883b925a0002bb880d6ec10/urdfModifiers/core/fixedOffsetModifier.py#L227-L229


Is it sound to you ?

**Comment by traversaro** [comment]
> Is it sound to you ?

Yes, thanks! The only fishy is:

> Note for computing the parent offset, only one joint is taken into account, namely the first one.

I am a bit confused on why this happens. Are we sure about the equation:
\begin{equation}
s_o = v_o- v_l \frac{sign(j_{o,1})}{2}
\end{equation}
? 

To be honest, I think I introduced (if I introduced it, I do not remember) that `sign(j_{o,1})` with the assumption that $sign(v_o) = sign(j_{o,1}) = sign(j_{o,2})$, and I do not know if this assumption is not respected. Probably we can just enforce this assumption in the code (i.e. raise an error if this is not the case) and so we solve the ambiguity of "preferring" a joint w.r.t. to another?

**Comment by CarlottaSartore** [comment]
In the end, I think the more robust solution would be to change the above formulation as follow: 

$$
\begin{equation}
e\_{o,i} = v\_o +v\_l \frac{sign(j\_{oi})}{2} - j\_{o,i}
\end{equation}
$$

$$
\begin{equation}
s\_o = (v\_o- \frac{v\_l}{2})sign(v\_o)
\end{equation}
$$

$$
\begin{equation}
v\_l^{'} = kv\_l
\end{equation}
$$

$$ 
\begin{equation}
v\_{o}^{'} = (s\_o+\frac{v\_l{'}}{2})sign(v\_0)
\end{equation}
$$

$$
j\_{o,i}^{'} = s\_o + v\_{l}^{'}sign({j\_{o,i}}) - e\_{o,i}
$$

This should work in different cases, I was studying the case in which $v\_o = 0 $ as per the attached representation
Untitled Diagram drawio (6)



Seems to me that everything works fine for the computations, indeed: 

$$
\begin{equation}
e\_{o,i} = v\_l \frac{sign(j\_{oi})}{2} - j\_{o,i}
\end{equation}
$$

$$
\begin{equation}
s\_o = (- \frac{v\_l}{2})sign(0)= - \frac{v\_l}{2}
\end{equation}
$$

$$
\begin{equation}
v\_l^{'} = kv\_l
\end{equation}
$$

$$ 
\begin{equation}
v\_{o}^{'} = - \frac{v\_l}{2}+v\_l{'}\frac{sign(0)}{2}= - \frac{v\_l}{2} + K\frac{v\_l}{2} = \frac{v\_l}{2} (k-1)
\end{equation}
$$

**Note** This would mean that $v\_o^{'}$ will not always be zero but it will be, such that $s\_o$ will be equal to $-\frac{v\_l}{2}$ always, as per the other cases. 

$$
j\_{o,i}^{'} = s\_o + v\_{l}^{'}sign({j\_{o,i}}) - e\_{o,i}
$$

Case in which the $v\_o$ is negative 
Untitled Diagram drawio (5)

Seems also that everything is working 

@traversaro  could you please double-check it? thanks :)  I will then code it and check also the tests!  

I think that with this implementation, we should get rid of the need for the following warning: 

> To be honest, I think I introduced (if I introduced it, I do not remember) that sign(j_{o,1}) with the assumption that sign(vo)=sign(jo,1)=sign(jo,2), and I do not know if this assumption is not respected. Probably we can just enforce this assumption in the code (i.e. raise an error if this is not the case) and so we solve the ambiguity of "preferring" a joint w.r.t. to another?

File: urdfModifiers/core/fixedOffsetModifier.py (hunk 1 above)
Comment (GrmanRodriguez) [reply urdfModifiers/core/fixedOffsetModifier.py]: Shouldn't we also add a test of the multiple joints feature itself? For this we need to also modify the urdf

**Comment by CarlottaSartore** [comment]
Gentle reminder @traversaro  :)

**Comment by traversaro** [comment]
> Gentle reminder @traversaro :)

Thanks, I had missed this.

**Comment by traversaro** [comment]
> @traversaro could you please double-check it? thanks :) I will then code it and check also the tests!

I am a bit confused by the definition of s_o (eq. 7), it seems discontinuous around v_o = 0, and that does not seems correct. I will try to write down the different cases, but if you are available we can discuss on the whiteboard.

**Comment by traversaro** [comment]
> > @traversaro could you please double-check it? thanks :) I will then code it and check also the tests!
> 
> I am a bit confused by the definition of s_o (eq. 7), it seems discontinuous around v_o = 0, and that does not seems correct. I will try to write down the different cases, but if you are available we can discuss on the whiteboard.

To clarify, this is the discontinuity: 
discontinuity


(plotted by plotting `(x-y/2)*sgn(x)` with https://c3d.libretexts.org/CalcPlot3D/index.html, with `x = v_o` and `y = v_l`).

I think anyhow I got it now, basically if `v_o` changes sign, the "side" of the box that is used to compute `s_o` changes, hence the discontinuity. In that case, it seems to be that everything is working fine, but probably the best way to check is to modify the test in https://github.com/icub-tech-iit/urdf-modifiers/blob/9e8fff10e4a6812defc1e9e3b1a1bb30ce2b480b/tests/tests.py#L868 to ensure that the offset indeed does not change for all the possible cases of relative signs between $v_o$ and $e_{o,i}$ .

File: tests/tests.py (hunk 2 above)
Comment (GrmanRodriguez) [reply tests/tests.py]: I would advise against using raw dictionaries in general. 

Next time we need to change the code we will need to remember what the dictionary is supposed to contain (and under which keys) as it's not self-evident.

From what I understand the idea of the dict is to be able to have each offset and know to which joint it corresponds. Maybe we can create a small `Offset` class, that has two properties: the joint (or joint name if that's better) and the offset number itself.

That way `parent_joint_offset` can be an instance of the `Offset` class, and `child_joint_offset` can be an array of objects of this type. Then next time we look at the code we know exactly what the function returns and it's easier to maintain.

Of course we'd need to also change the functions that call this function to accept the new class and use it as intended.

What do you think @AlexAntn ?

**Comment by CarlottaSartore** [comment]
I have updated the definition of all the variables as per the comment https://github.com/icub-tech-iit/urdf-modifiers/pull/24#issuecomment-1081834443 and rebase on top of master. 

All the tests are passing! @traversaro

File: tests/tests.py (hunk 2 above)
Comment (AlexAntn) [reply tests/tests.py]: I agree a class would work better also on the long term, dicts tend to be hard to manage. What do you think @CarlottaSartore ?

**Review by traversaro — APPROVED** [review]

**Comment by traversaro** [comment]
> I have updated the definition of all the variables as per the comment [#24 (comment)](https://github.com/icub-tech-iit/urdf-modifiers/pull/24#issuecomment-1081834443) and rebase on top of master.
> 
> All the tests are passing! @traversaro

Thanks, good for me, even if I guess there are still comments from the others.

File: tests/tests.py (hunk 2 above)
Comment (GrmanRodriguez) [reply tests/tests.py]: Changes implemented in [this](https://github.com/icub-tech-iit/urdf-modifiers/pull/24/commits/26668f313fef0775bd1801ae4440facab6093529) commit @CarlottaSartore @AlexAntn

**Pull request merged**